node/src/effect.rs

```diff
@@ -1489,16 +1489,51 @@ impl<REv> EffectBuilder<REv> {
         .await
     }
 
-    /// Check if validator is bonded in the era.
-    pub(crate) async fn is_bonded_validator(self, era_id: EraId, public_key: PublicKey) -> bool
-    where
-        REv: From<ConsensusRequest>,
-    {
-        self.make_request(
-            |responder| ConsensusRequest::IsBondedValidator(era_id, public_key, responder),
-            QueueKind::Regular,
-        )
-        .await
+    /// Checks whether the given validator is bonded in the given era.
+    pub(crate) async fn is_bonded_validator(
+        self,
+        validator: &PublicKey,
+        era_id: EraId,
+        activation_era_id: EraId,
+        initial_state_root_hash: Digest,
+        protocol_version: ProtocolVersion,
+    ) -> Option<bool>
+    where
+        REv: From<ContractRuntimeRequest> + From<StorageRequest>,
+    {
+        if era_id == activation_era_id {
+            // in the activation era, we read the validators from the global state; we use the
+            // global state hash of the first block in the era, if it exists - if we can't get it,
+            // we use the initial_state_root_hash passed from the chainspec loader
+            let maybe_key_block = self.get_key_block_for_era_id_from_storage(era_id).await;
+            let root_hash = match maybe_key_block {
+                None => initial_state_root_hash,
+                Some(key_block) => self
+                    .get_block_at_height_from_storage(key_block.height() + 1)
+                    .await
+                    .map(|block| *block.state_root_hash())
+                    .unwrap_or(initial_state_root_hash),
+            };
```

Comment:
I think we can always use a state root hash from the last block from the linear chain we know of and use that for the query. Contract Runtime should know about past era validators (up to the unbonding period). I will confirm that and will get back to you b/c it would simplify this thing a lot and would let us remove all the new arguments from the constructor of the `LinearChain`.

Reply:
It has to remember who was the validator b/c of the cross-era slashing, right?

Reply:
OK, it _can_ slash for past eras but it (currently) cannot answer the request who was the validator in some past era.